### PR TITLE
Refactored atoms to allow more funcionality

### LINKED ISSUE #9: req: Allow access to internal service
Is it possible for services to be exposed or accessible.

This allow us to hook up listeners post initialization.

### PR DESCRIPTION
**Pull request by brunocangs**
This PR breaks down the `useMachineAtom` function, separating it's internal functionalities in order to make them usable separately. 

This PR closes #9 by introducing the `atomWithActor` function. It takes care of creating the Actor, based on the provided Logic and options. The type definitions correctly indicate when the Logic requires `inputs`, such as in the example added. It's setter function works exactly like the setter function for `atomWithMachine`

It also adds a `atomWithActorSnapshot` function, that handles the subscription logic to maintain the latest snapshot in state. Its setter function allows you to reset the subscription and clear its state whenever needed.

There is a type definition I had to bruteforce with a `as any` even though the input is correct and using the same type as their internal definitions it would always error for me

### CODE REVIEW

File: examples/01_typescript/src/app.tsx [hunk 1]
```diff
@@ -3,8 +3,12 @@
...
 import { atom } from 'jotai/vanilla';
 // TODO: Revert back to regular import
-import { RESTART, atomWithMachine } from '../../../src/atomWithMachine';
-import { assign, setup } from 'xstate';
+import { useEffect } from 'react';
+import { assign, fromPromise, setup } from 'xstate';
+import { atomWithActor } from '../../../src/atomWithActor';
+import { atomWithActorSnapshot } from '../../../src/atomWithActorSnapshot';
+import { atomWithMachine } from '../../../src/atomWithMachine';
+import { RESTART } from '../../../src/utils';
```
Comment (dai-shi) [inline examples/01_typescript/src/app.tsx]: I think you can import from `jotai-xstate`.
It makes the example standalone, and works outside the repo.

File: src/index.ts [hunk 2]
```diff
@@ -1 +1,5 @@
-export { atomWithMachine, RESTART } from './atomWithMachine.js';
+export { atomWithMachine } from './atomWithMachine.js';
+export { isGetter, RESTART } from './utils.js';
+export type { Gettable } from './utils.js';
```
Comment (dai-shi) [inline src/index.ts]: Can we avoid exporting `isGetter` and `Gettable`. They feel internal things.
```suggestion
export { RESTART } from './utils.js';
```

**Review by dai-shi — COMMENTED** [review]
Thanks for maintaining the library!!

I hope someone can review the core part, but I'm afraid no one is available.

Here's some minor comments of mine.

File: examples/01_typescript/src/app.tsx (hunk 1 above)
Comment (brunocangs) [reply examples/01_typescript/src/app.tsx]: Does it link correctly before the update is published? I will give a try

Comment (dai-shi) [reply examples/01_typescript/src/app.tsx]: it should be handled by tsconfig alias.

**Comment by brunocangs** [comment]
@dai-shi I fixed the requested changes. Didn't realize there was a command to run the example with the vitest config, that's why it was failing for me 😄

**Comment by dai-shi** [comment]
> a command to run the example with the vitest config

yeah, that was it.

File: examples/01_typescript/src/app.tsx [hunk 3]
```diff
@@ -2,9 +2,14 @@
...
-import { RESTART, atomWithMachine } from '../../../src/atomWithMachine';
-import { assign, setup } from 'xstate';
+import {
+  atomWithActor,
+  atomWithActorSnapshot,
+  atomWithMachine,
+} from 'jotai-xstate';
+import { useEffect } from 'react';
+import { assign, fromPromise, setup } from 'xstate';
+import { RESTART } from '../../../src/utils';
```
Comment (dai-shi) [inline examples/01_typescript/src/app.tsx]: This should also be imported from 'jotai-xstate'.

**Review by dai-shi — APPROVED** [review]
This looks overall good to me.

But as I don't understand every details, it feels adding tests would be great and make maintainable.
Would you like to work on tests in this PR or in a separate PR?

**Comment by brunocangs** [comment]
> This looks overall good to me.
> 
> But as I don't understand every details, it feels adding tests would be great and make maintainable. Would you like to work on tests in this PR or in a separate PR?

I feel like both tests and documentation are important for these changes. Either works for me, but I'm not sure how soon I'll have the bandwith to work on it. Possibly only in a few weeks

**Comment by dai-shi** [comment]
@brunocangs ?

**Comment by brunocangs** [comment]
> @brunocangs ?

Hi,

I'm sorry for leaving this hanging for so long, I've added some basic tests to the new atoms, testing with a promise and testing listening to its events. Let me know if there are other tests you would like to see implemente

**Comment by dai-shi** [comment]
Hey, thanks for working on it.
I think we should ship it. Would you like to do it?
- Merge this PR
- Update CHANGELOG.md and update package version
- Tag version and push it. (`git tag v0.6.0`)

**Comment by brunocangs** [comment]
@dai-shi also added some documentation and incremental examples for the new methods on the readme. Please let me know if you would like any changes.

If not I'll merge, tag and release it

**Review by dai-shi — APPROVED** [review]
LGTM (I really can't review from XState perspective though.)

Sure, go ahead.

**Pull request merged**